base/libgit2/callbacks.jl

```diff
@@ -31,74 +31,131 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
+    isusedcreds = isused(creds)
+
+    # use ssh key or ssh-agent
+    if isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_KEY))
+        credid = "ssh://$host"
+        # first try ssh-agent if credentials support its usage
+        if isdefined(creds, :use_ssh_agent) && getfield(creds, :use_ssh_agent) == 0x00
+            err = ccall((:git_cred_ssh_key_from_agent, :libgit2), Cint,
+                         (Ptr{Ptr{Void}}, Cstring), cred, username_ptr)
+            setfield!(creds, :use_ssh_agent, 0x01) # used ssh-agent only one time
+            err == 0 && return Cint(0)
         end
 
-        isempty(username) && isempty(userpass) && return Cint(-1)
-
-        err = ccall((:git_cred_userpass_plaintext_new, :libgit2), Cint,
-                     (Ptr{Ptr{Void}}, Cstring, Cstring),
-                     cred, username, userpass)
-        err == 0 && return Cint(0)
-    elseif isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_KEY)) && err > 0
-        # use ssh-agent
-        err = ccall((:git_cred_ssh_key_from_agent, :libgit2), Cint,
-                     (Ptr{Ptr{Void}}, Cstring), cred, username_ptr)
-        err == 0 && return Cint(0)
-    elseif isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_CUSTOM)) && err > 0
-        # for SSH we need key info, look for environment vars SSH_* as well
+        errcls, errmsg = Error.last_error()
+        if errcls != Error.None
+            # Check if we used ssh-agent
+            if isdefined(creds, :use_ssh_agent) && getfield(creds, :use_ssh_agent) == 0x01
+                println("ERROR: $errmsg ssh-agent")
+                setfield!(creds, :use_ssh_agent, 0x02) # reported ssh-agent error
+            else
+                println("ERROR: $errmsg")
+            end
+        end
 
         # if username is not provided, then prompt for it
-        username = if username_ptr == Cstring_NULL
-            prompt("Username for '$url'")
+        username = if username_ptr == Cstring(C_NULL)
+            uname = creds[:user, credid] # check if credentials were already used
+            uname !== nothing && !isusedcreds ? uname : prompt("Username for '$schema$host'")
         else
             bytestring(username_ptr)
         end
+        creds[:user, credid] = username # save credentials
 
-        publickey = if "SSH_PUB_KEY" in keys(ENV)
-            ENV["GITHUB_PUB_KEY"]
+        # For SSH we need key info, look for environment vars SSH_* as well
+        publickey = if haskey(ENV,"SSH_PUB_KEY")
+            ENV["SSH_PUB_KEY"]
         else
-            keydef = homedir()*"/.ssh/id_rsa.pub"
-            prompt("Public key location", default=keydef)
+            keydef = creds[:pubkey, credid] # check if credentials were already used
+            if keydef !== nothing && !isusedcreds
+                keydef # use cached value
+            else
+                keydef = homedir()*"/.ssh/id_rsa.pub"
+                prompt("Public key location for '$schema$username@$host'", default=keydef)
+            end
```

Comment:
this "`hasenv` otherwise `prompt`" pattern looks like it should be refactored into a common function